File: frontend/src/types/workflow.ts
export type WorkflowStatus = 'pending' | 'running' | 'completed' | 'failed';

export type AgentRole =
  | 'planner'
  | 'researcher'
  | 'technical_architect'
  | 'developer'
  | 'reviewer'
  | 'final_answer';

export type WorkflowEventType =
  | 'workflow_started'
  | 'agent_pending'
  | 'agent_running'
  | 'agent_completed'
  | 'agent_failed'
  | 'workflow_completed'
  | 'workflow_failed';

export type WorkflowValue = string | Record<string, unknown> | null;

export interface HealthResponse {
  status: string;
  version: string;
  environment: string;
}

export interface WorkflowStep {
  id: string;
  role: AgentRole;
  name: string;
  description?: string | null;
  input?: WorkflowValue;
  output?: WorkflowValue;
  status: WorkflowStatus;
  error?: string | null;
  created_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
}

export interface WorkflowRun {
  id: string;
  input?: WorkflowValue;
  task?: string;
  output?: WorkflowValue;
  status: WorkflowStatus;
  steps: WorkflowStep[];
  final_answer?: string | null;
  error?: string | null;
  created_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
  total_duration_ms?: number | null;
}

export interface WorkflowEvent {
  id: string;
  workflow_id: string;
  event: WorkflowEventType;
  status?: WorkflowStatus | null;
  role?: AgentRole | null;
  step?: WorkflowStep | null;
  workflow?: WorkflowRun | null;
  message?: string | null;
  created_at: string;
}

export interface RunWorkflowRequest {
  task: string;
  workflow_id?: string;
}

File: backend/app/models/workflow.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import AgentRole, WorkflowStatus


class WorkflowStep(BaseModel):
    """One agent step inside an orchestration workflow."""

    model_config = ConfigDict(str_strip_whitespace=True)

    id: str = Field(default_factory=lambda: str(uuid4()))
    role: AgentRole
    name: str = Field(min_length=1)
    description: str | None = Field(default=None, min_length=1)
    input: str | dict[str, Any] | None = None
    output: str | dict[str, Any] | None = None
    status: WorkflowStatus = WorkflowStatus.PENDING
    error: str | None = None
    created_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = Field(default=None, ge=0)


class WorkflowRun(BaseModel):
    """Runtime representation of a full multi-agent workflow execution."""

    model_config = ConfigDict(str_strip_whitespace=True)

    id: str = Field(default_factory=lambda: str(uuid4()))
    input: str | dict[str, Any] = Field(min_length=1)
    status: WorkflowStatus = WorkflowStatus.PENDING
    steps: list[WorkflowStep] = Field(default_factory=list)
    error: str | None = None
    created_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = Field(default=None, ge=0)


class WorkflowResult(BaseModel):
    """Final result returned after a workflow finishes."""

    id: str = Field(default_factory=lambda: str(uuid4()))
    input: str | dict[str, Any] | None = None
    output: str | dict[str, Any] | None = None
    final_answer: str | None = None
    status: WorkflowStatus
    steps: list[WorkflowStep] = Field(default_factory=list)
    error: str | None = None
    created_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = Field(default=None, ge=0)
    total_duration_ms: int | None = Field(default=None, ge=0)


__all__ = [
    "WorkflowResult",
    "WorkflowRun",
    "WorkflowStatus",
    "WorkflowStep",
]
